src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -843,8 +843,10 @@ private EnumeratorResult GetEnumeratorInfo(ref ForEachEnumeratorInfo.Builder bui
                         builder.GetEnumeratorMethod.ReturnType.SpecialType == SpecialType.System_Collections_IEnumerator);
                 }
 
-                // We don't know the runtime type, so we will have to insert a runtime check for IDisposable (with a conditional call to IDisposable.Dispose).
+                // We don't know the runtime type, so we will have to insert a runtime check for disposal
+                // that will call the interface method if the check is successful
                 builder.NeedsDisposal = true;
+                builder.DisposeMethod = this.Compilation.GetWellKnownDisposeMethod(isAsync);
```

Comment:
>this.Compilation.GetWellKnownDisposeMethod(isAsync); [](start = 40, length = 52)

We are locating a method without reporting any use site diagnostics. Why do we need to do this? I think there is no real need for us to pass method symbols for non-pattern-base disposal. Flow graph builder knows how to find them. So I wouldn't expect us to make any changes related to non-pattern-based disposal

Reply:
The idea behind this change was the CFG wouldn't do the finding of the method. Instead we make all the logic decisions in the binder once, and CFG/IOperation just reports what was found during binding. If we keep finding the method later on we end up smearing out the logic in multiple places, and end up with bugs like the one this PR is aiming to fix, where the logic isn't updated in all the various places it's used.

Reply:
> and end up with bugs like the one this PR is aiming to fix, where the logic isn't updated in all the various places it's used.

I do not believe this is the case. It is enough for us to record the method that was found based on a pattern. We can deal interface based disposal the same way we did before. A null value is good enough to represent the scenario and all the interested parties know how to deal with interface-based disposal. I do not see any good reason to make any changes around these scenarios. The risk to break things is high and the benefit is very small in my opinion.  

---
In reply to: [533647453](https://github.com/dotnet/roslyn/pull/49481#discussion_r533647453) [](ancestors = 533647453)